File: results/RQ1/05_baseline_qualification/units/p06_tank_battle_game/java/replays/0001_seed_573870900/p06_tank_battle_game_java/p06_tank_battle_game_java/src/main/java/Arena.java
/**
 * Arena bounds and obstacles with validation and defensive copy.
 */
public class Arena {
    private final int width;
    private final int height;
    private final ObstacleMap obstacles;

    /**
     * Creates a new arena with the given dimensions and obstacles.
     *
     * @param width the width of the arena
     * @param height the height of the arena
     * @param obstacles the obstacle map for this arena
     * @throws IllegalArgumentException if width or height is not positive,
     *                                  or if obstacles is null
     */
    public Arena(int width, int height, ObstacleMap obstacles) {
        if (width <= 0) {
            throw new IllegalArgumentException("Arena width must be positive");
        }
        if (height <= 0) {
            throw new IllegalArgumentException("Arena height must be positive");
        }
        if (obstacles == null) {
            throw new IllegalArgumentException("Arena obstacles cannot be null");
        }

        this.width = width;
        this.height = height;
        this.obstacles = obstacles.copy();
    }

    /**
     * Checks if the given position is within the bounds of this arena.
     *
     * @param p the position to check
     * @return true if the position is within bounds, false otherwise
     */
    public boolean inBounds(Position p) {
        return p.getX() >= 0 && p.getX() < width && p.getY() >= 0 && p.getY() < height;
    }

    /**
     * Checks if there is an obstacle at the given position.
     *
     * @param p the position to check
     * @return true if there is an obstacle at the position, false otherwise
     */
    public boolean isObstacle(Position p) {
        return obstacles.isBlocked(p);
    }

    /**
     * Creates a copy of this arena.
     *
     * @return a new arena with the same dimensions and obstacles
     */
    public Arena copy() {
        return new Arena(width, height, obstacles);
    }
}

File: results/RQ1/05_baseline_qualification/units/p06_tank_battle_game/python/replays/0001_seed_2408978515/p06_tank_battle_game_python/p06_tank_battle_game_python/tank_battle/arena.py
"""Arena bounds and blocking rules."""

from .obstacle import Obstacle
from .geometry import Rect


class Arena:
    """Represents a bounded arena with obstacles.

    Attributes:
        width: The width of the arena.
        height: The height of the arena.
        obstacles: A list of obstacles in the arena.
    """

    def __init__(self, width: int, height: int, obstacles: list[Obstacle]):
        """Initialize an arena with given dimensions and obstacles.

        Args:
            width: The width of the arena (must be positive).
            height: The height of the arena (must be positive).
            obstacles: A list of obstacles in the arena.

        Raises:
            ValueError: If width or height is not positive.
        """
        if width <= 0:
            raise ValueError("Width must be positive")
        if height <= 0:
            raise ValueError("Height must be positive")

        self.width = width
        self.height = height
        self.obstacles = obstacles

    def in_bounds(self, x: int, y: int) -> bool:
        """Check if a position is within the arena bounds.

        Args:
            x: The x-coordinate to check.
            y: The y-coordinate to check.

        Returns:
            True if the position is within bounds, False otherwise.
        """
        return 0 <= x < self.width and 0 <= y < self.height

    def is_blocked(self, x: int, y: int) -> bool:
        """Check if a position is blocked by arena boundaries or obstacles.

        A position is blocked if it's outside the arena or occupied by an obstacle.

        Args:
            x: The x-coordinate to check.
            y: The y-coordinate to check.

        Returns:
            True if the position is blocked, False otherwise.
        """
        # First check if position is out of bounds
        if not self.in_bounds(x, y):
            return True

        # Then check if position is occupied by any obstacle
        for obstacle in self.obstacles:
            if obstacle.rect.contains(x, y):
                return True

        return False